README.md
# Text Game
This is a work-in-progress text-based game I am making. Although there is no story yet, the game should work if you download the dependencies listed below.
## Dependencies
You will need:
  1. simpleaudio (`pip install simpleaudio`)
  2. playsound (`pip install playsound`)
  3. progbar (`pip install -i https://test.pypi.org/simple/ progbar`)
  
The last one is my own custom module.


change_log.txt
CHANGELOG ONLY RECORDS CHANGES MADE POST 1/3/2020

1/4/2020:
    Enemy items:
        Added Enemy_Item Class to systems.py:
            # still need to add 'special' compatability
            # works almost the same as player's item usage
        Added 'inventory' key to enemy dict template in story.py
            # needs to be added to current enemies
    Other:
        created change_log.txt
1/25/2020:
    Enemy Items:
        Completed Enemy_Item Class:
            # added 'special' compatability
        Added Enemy_Item functionality to battle in systems.py:
    Functions:
        Added configure_params function to systems.py:
            # configures params for 'special' functions
    Other:
        Added more comments
        Moved save files to saves
2/7/2020:
    Items:
        Added universal Item class:
            # standardizes __init__ for Enemy_Item and Player_Item
        Added Player_Item class:
            # works almost identically as Enemy_Item
            # tested extensively; works perfectly
    Functions:
        Changed add_item to be compatible with Player_Item
        Modified battle to work with new Player_Item class
    Other:
        Added debug area to battle
        Improved readablilty and practicality of code
2/12/2020:
    Functions:
        Added speed_factor to text_scroll
        Fixed voice files in text_scroll
        Made turn_based_damage compatible for enemy use
    Other:
        Uploaded text_game to github (https://github.com/dirkwind/text-game)
        Added README.md
        Added comments explaining most files' purpose
2/19/2020:
    Functions:
        Changed the way special functions are called
            # now uses function(*[param_list]) rather than function([param_list])
2/28/2020:
    Functions:
        text_scroll
            # improved the handling of voice_file if it is None or not
            # added comments and doctrings
        quicktime_bar
            # added default value for key ('f')
        All
            # added comments and doctrings
    Other:
        .gitignore
            # finally got it to ignore files
3/25/2020:
    Functions:
        ana_bomb
            # fixed unused variable error
    Entities:
        Player:
            # added player class
            # changed most player value calls to class format (player['health'] --> player.health)
        Entity:
            # added Entity class
            # added __getitem__ and __setitem__ methods
                - dict value call format (e.g. player['health']) still works for getting and setting values within entity subclasses
9/19/2020:
    Items:
        Item superclass:
            # added special, heal, damage, and stat_change classmethods
                - makes creating with each pupose easier/simpler
    Entities:
        Player:
            # added add_item method
                - replaces the add_item function. Adds an item to the player's inventory
            # added attacked method
                - replaces the player_attacked function, does the same thing 
        Enemy:
            # added add_item method
                - replaces the add_item function. Adds an item to an enemy's inventory
            # added attacked method
                - replaces the contents of choice 1 in the battle function, does the same thing
    Story:
        Enemies:
            # fixed a typo in the doge (raptor) enemy
            # changed 'Hamburger' Item to use new method of creating items
        Other:
            # changed 'Sketchy Throwing Knife' Player_Item creation to new method

story.py
# this is where the story is written

from systems import *
from aux_functions import *
from playsound import *
from time import *

# check enemy_format.txt for enemy formatting info

dog = Enemy.from_dict({
    'name': 'Gremlin',
    'intro': 'You\'ve encontered a Gremlin!', # intro message
    'description': 'A fat couch potato.', # short description of enemy
    'checkable': [True, "name can't be checked."], # whether or not player can check enemy's stats
    'attack': 1,
    'defense': 10,
    'speed': 4, # 4+ = no Great Hits, 8+ = no Good Hits, 13+ = no Fair Hits, 20+ = 100% immune to damage items, all keep Critical (nums in b/w included)
    'health': 20,
    'max_health': 20,
    'give_up': [0, 'Blarg... You Win!'],
    'can_flee': False,
    'bonus_xp': random.randint(1,40), # extra xp for the player if they defeat this enemy
    'inventory': [ # Enemy_Item(self, name, purpose='heal, stat_bonus, damage', value, uses, usage_chance, dodgable=None, turns=0) 
        Enemy_Item('Goblin Engergy Jab', 'special', None, 3, False, special_func=turn_based_damage, special_params=['$player', '$enemy', '$turns', 1, True])
    ],
    'use_item_chance': 20, # between and including 1 through 100; percentage chance (w/ precision of 3 decimal places) that the enemy will use an item
    'responses': { # use (['', ''], 0, stat, value) to change a stat's value when an act is performed
        'threaten': [
            (['"I destroy your hentai!"', 'Gremlin tenses up'], 20, 'defense', 2),
            (['"Oh what a weeb, what a weeb!"', '"Are you trying to make me laugh by insulting me and my tradition from afar?"'], 20)
            ],
        'be a chad': [
            (['You beat your chest and grunt repeatedly.', '"Are you making fun of me?"'], -5)
            ],
        'rub belly': [
            (['You carress his massive stomach.', '"Giggity giggity goo!"'], 30)
            ],
        'indulge': [
            (['You read manga with the foul creature.', 'Mmm...'], 100)
            ]
    }
})

doge = Enemy.from_dict({
    'name': 'Raptor',
    'intro': '"rawr"', # intro message
    'description': 'Just a velociraptor.', # short description of enemy
    'checkable': [True, "name can't be checked."], # whether or not player can check enemy's stats
    'attack': 7,
    'defense': 4,
    'speed': 12, # 4+ = no Great Hits, 8+ = no Good Hits, 13+ = no Fair Hits, 20+ = 100% immune to damage items, all keep Critical (nums in b/w included)
    'health': 15,
    'max_health': 15,
    'give_up': [player['level'] * 10, "rawr..."],
    'can_flee': True,
    'bonus_xp': 0, # extra xp for the player if they defeat this enemy
    'inventory': [ # Enemy_Item(self, name, purpose='heal, stat_bonus, damage', value, uses, usage_chance, dodgable=None, turns=0) 
        #Enemy_Item('Hamburger', 'stat_change', 10, 1, turns=3, changed_stat='attack')
        Enemy_Item.stat_change("Hamburger", 10, 1, 3, 'attack')
    ],
    'use_item_chance': 5, # between and including 1 through 100; percentage chance (w/ precision of 3 decimal places) that the enemy will use an item
    'responses': { # use (['', ''], 0, stat, value) to change a stat's value when an act is performed
        'feed meat': [
            (['You toss a leg of chicken in the air...', 'rawr'], 100)  
            ],
        'tame': [
            (['You hold your hand out.', 'Raptor touches you.'], 50)
            ],
        'stand still': [
            (['You stand completely still...', 'Raptor couldn\'t see you so it ran away.'], 100)
            ],
        'shoo': [
            (['You tell it to leave...', 'Raptor leaves...'], 100)
            ]
    }
}
)

#player.add_item(Player_Item('Anaesthetic Bomb', 'special', None, 2, special_func=ana_bomb, special_params=['$enemy', 1]))
#player.add_item(Player_Item('Effect Longevity Boost', 'special', None, 3, special_func=change_turns, special_params=['$turns', -4, '&turns']))
player.add_item(Player_Item.special('Sketchy Throwing Knife', 3, vampire, ['$enemy', '$player', 3]))
#player.add_item(Player_Item('Charge Attack', 'special', None, 2, special_func=turn_based_damage, special_params=['$enemy', '$turns', 0.45, True]))

battle(doge)

aux_functions.py
# this is where special item functions go

import json
import math
import msvcrt
import random
import sys
import time

from systems import *

def ana_bomb(target: dict):
    '''Sets targeted entity's speed to 0 unless its speed is <= 0. params=[target]
    NOTE: cannot be used on players because they have no 'speed' value
    '''
    if target["speed"] > 0:
        target["speed"] = 0
        text_scroll(f'\n{target["name"]} was anaesthetized! Their SPEED is now 0!\n')
    else:
        text_scroll(f'\n{target["name"]} was too slow for the anaesthesia to take effect!\n')

def turn_based_damage(target: dict, user: dict, turn_number: int, damage_factor: float, dodgable: bool):
    '''Attack that bases its damage on the current turn. params=[target, user, '$turns', damage_factor, dodgable]'''

    if target is player and dodgable:
        text_scroll("\nGet ready to dodge!\n")
        time.sleep(1)
        chance_num = int(100 - ((quicktime_bar('f', user['speed']) / 30) * 100))
        speed = user['speed']
    else:
        chance_num = random.randint(1, 10000)/100
        speed = target['speed']

    if speed == 20:
        percent = 100
    else:
        percent = -100 * (((speed/20)-1)**2) + 100

    if chance_num <= percent or not dodgable:
        damage = int((user['attack'] * turn_number * damage_factor)) - target['defense']
        if damage <= 0: damage = 0
        target['health'] -= damage
        text_scroll(f'\n{"You" if target is player else target["name"]} took {damage} damage! {"You" if target is player else "They"} are now at {target["health"]} HP!\n')
    else:
        text_scroll(f'\n{"You" if target is player else target["name"]} dodged the attack!\n')


def change_turns(turn_number: int, change_amount: int, clarifier="&turns"):
    '''Changes the value of turns params=['$turns', change, '&turns']'''
    text_scroll(f'\nActive effects will now last for {-change_amount} more turns!\n')
    return turn_number + change_amount

def vampire(target: dict, user: dict, divisor: float):
    '''Life steal. params=[target, user, divisor]
    divisor determines lifesteal (health_increment = damage // divisor)
    '''
    damage = random.randint(int(user['attack'] * 0.8), int(user['attack'] * 0.9)) - target['defense']
    if damage <= 0:
        text_scroll('\nThe attack did 0 damage!\n')
    else:
        text_scroll(f'\nThe attack did {damage} damage to {"you" if target == player else target["name"]}! {"You" if user == player else user["name"]} gained {int(damage // divisor)} HP due to LIFE STEAL!\n')
        target['health'] -= damage
        user['health'] += int(damage // divisor)


systems_backup.py
import math
import msvcrt # used solely in quicktime events
import random
import sys
import time
import shelve # used for saving data

import simpleaudio as sa

sys.path.append('../')

import progbar # quicktime event visualizer

class Enemy_Item(Item):
    
    active_items = []

    def __init__(self, name: str, purpose: str, value: int, uses: int, silent=False, **unique_args):
        '''purpose is 'stat_change', 'heal', 'damage', or 'special
        use dodgable=[True/False] if purpose = 'damage'
        use turns=[num_of_turns: int] and changed_stat=[stat_name: str] if purpose = 'stat_change'
        use special_func=[function] and special_params=[function_params: list/tuple] if purpose = 'special'
        '''

        self.name = name
        self.purpose = purpose
        self.value = value
        self.uses = uses
        self.silent = silent
        self.special_params = []
        if purpose == 'stat_change':
            self.turns = unique_args['turns']
            self.changed_stat = unique_args['changed_stat']
        elif purpose == 'damage':
            self.dodgable = unique_args['dodgable']
        elif purpose == 'special':
            self.special_func = unique_args['special_func']
            self.special_params = unique_args['special_params']
    
    @classmethod
    def clear_active(cls):
        cls.active_items.clear()

    @classmethod
    def check_active(cls, current_turn, enemy):
        for item in cls.active_items:
            # index 0 : purpose, 1 : value, 2 : turn_when_deactivates, 3 : item_name
            if current_turn >= item[2]:
                enemy[item[0]] -= item[1]
                text_scroll(f'\n{enemy["name"]}\'s item, {item[3].upper()}, has worn off!\n')


    def activate(self, current_turn=0, enemy={}, bonus_def=0):
        # checks if an item should have been used
        if self.uses <= 0: 
            print('Error: Enemy_Item used with {} uses left'.format(self.turns))
            return None
        self.uses -= 1
        
        # prints a message stating that an enemy used an item
        if not self.silent:
            text_scroll(f'\n{enemy["name"]} used {self.name.upper()}!\n')
        
        # item functions
        if self.purpose == 'heal':
            enemy['health'] += self.value

            if enemy['health'] > enemy['max_health']:
                enemy['health'] = enemy['max_health']
                if not self.silent:
                    text_scroll(f'\n{enemy["name"]} reached full HP!\n')
            else:
                if not self.silent:
                    text_scroll(f'\n{enemy["name"]} recovered {self.value} HP!\n')

        elif self.purpose == 'stat_change':
            enemy[self.changed_stat] += self.value
            Enemy_Item.active_items.append([self.changed_stat, self.value, self.turns + current_turn, self.name])
            if not self.silent:
                text_scroll(f'\n{enemy["name"]}\'s {self.changed_stat.upper()} has been increased by {self.value} for {self.turns} turns!\n')

        elif self.purpose == 'damage':

            if self.dodgable: # if the item can be dodged by the player
                text_scroll(f"\nGet ready to dodge the {self.name.upper()}!\n")
                base_attack = self.value
                time.sleep(1)
                difference = quicktime_bar('f', enemy['speed'])

                if difference == 0:
                    print('Dodged!')
                    attack = 0
                elif difference <= 2:
                    print('Grazed!')
                    attack = base_attack - math.floor(base_attack*0.3)
                elif difference <= 5:
                    print('Hit!')
                    attack = base_attack - math.floor(base_attack*0.085)
                elif difference <= 15:
                    print('Solid Hit!')
                    attack = base_attack
                else:
                    print('Critical!')
                    attack = base_attack + round(base_attack*0.25)

                if attack != 0:
                    damage = attack - (player['defense'] + bonus_def)
                    if damage < 0: damage = 0
                    player['health'] -= damage
                    text_scroll(f'\nYou were hit for {damage} damage! You are now at {player["health"]} HP!\n')
                else:
                    text_scroll(f'\nThe {self.name.upper()} missed!\n')

            else: # if it is impossible for the player to dodge the item
                damage = self.value - (player['defense'] + bonus_def)
                if damage < 0: damage = 0
                player['health'] -= damage
                text_scroll(f'\nYou couldn\'t avoid the {self.name.upper()}!\n')
                text_scroll(f'\nYou were hit for {damage} damage! You are now at {player["health"]} HP!\n')
        
        elif self.purpose == 'special':
            # returned in case special_func returns a value
            return self.special_func(configure_params(self.special_params, enemy, current_turn, self.silent))

def text_scroll(text, period=0.25, comma=0.15, normal=0.03, space=None, voice_file=None):
    if voice_file != None:
        wave_obj = sa.WaveObject.from_wave_file("sounds/voice_sans.wav")
    if space == None:
        space = normal
    for letter in text:
        sys.stdout.write(letter)
        sys.stdout.flush()
        if voice_file != None:
            wave_obj.play()
        if letter in set('.?!'):
            time.sleep(period)
        elif letter in set(',;:'):
            time.sleep(comma)
        elif letter == ' ':
            time.sleep(space)
        else:
            time.sleep(normal)
        

def quicktime_bar(key, speed_factor=0):
    percent = 0
    total = 60
    attack_bar.startProgress(f'Press {key.upper()}:')
    while percent < total:
        percent += 1
        inc_factor = lambda x : round(x * ((0.95) ** speed_factor))
        time.sleep(random.randint(inc_factor(10), inc_factor(20))/ 1000)
        attack_bar.updateProgress( attack_bar.toPercent( round(-2 * abs(percent - 30) + total ), total) )
        if msvcrt.kbhit():
            if ord(msvcrt.getch()) == ord(key):
                break
    attack_bar.pauseProgress()
    return abs(percent - 30)

def player_attacked(enemy, bonus_def):
    text_scroll("\nGet ready to dodge!\n")
    base_attack = enemy['attack']
    defense = player['defense']
    time.sleep(1)
    difference = quicktime_bar('f', enemy['speed'])
    if difference == 0:
        print('Dodged!')
        attack = 0
    elif difference <=2:
        print('Grazed!')
        attack = base_attack - math.floor(base_attack*0.4)
    elif difference <= 5:
        print('Hit!')
        attack = base_attack - math.floor(base_attack*0.1)
    elif difference <= 15:
        print('Solid Hit!')
        attack = base_attack
    else:
        print('Critical!')
        attack = base_attack + round(base_attack*0.2)
    if attack != 0:
        damage = attack - (defense + bonus_def)
        if damage < 0: damage = 0
        player['health'] -= damage
        text_scroll(f'\nYou were hit for {damage} damage! You are now at {player["health"]} HP!\n')
    else:
        text_scroll(f'\n{enemy["name"]} missed!\n')
        time.sleep(0.5)

def add_item(item='item name or [purpose, name, value, turns/dodgable/params, uses]', purpose='purpose', item_value=0, turns=1, uses=1, silent=False):
    '''Add an item to the player's inventory. purpose='heal', 'bonus attack', 'bonus defense', 'damage', 'special' '''
    added = False
    if type(item) != list:
        item_name = item
        item = [item, purpose, item_value, turns, uses]
    else:
        item_name = item[1]
    for itm in player['inventory']:
        if itm[:4] == item[:4]:
            itm[4] += item[4]
            added = True
    if len(player['inventory']) < 8:
        if not added: player['inventory'].append(item)
        if not silent:
            text_scroll(f'\n{item_name.upper()} has been added to your inventory!\n')
    else:
        if not silent:
            text_scroll(f'\nYour inventory is full.\n')

def game_over():
    sys.exit()

def gain_xp(enemy, silent=False):
    stat_mod = player['level'] // 5 + 1
    if type(enemy) != int:
        xp_gain = enemy["max_health"] + enemy["attack"] + enemy["defense"] + enemy["speed"] + enemy["bonus_xp"]
    else:
        xp_gain = enemy
    player['xp'] += xp_gain
    while player['xp'] >= player['xp_to_next']: 
        player['level'] += 1
        for stat in ['attack', 'defense']:
            player[stat] += 2 * stat_mod
        player['max_health'] += 7 * stat_mod
        player['health'] = player['max_health']
        player['xp_to_next'] = round(player['xp_to_next'] * 1.8)
        if not silent:
            text_scroll(f'\nYou LEVELED UP to LEVEL {player["level"]}!\n')
    else:
        if not silent:
            text_scroll(f'\nYou gained {xp_gain} XP! You need {player["xp_to_next"] - player["xp"]} more XP to level up!\n')

def check_win_conditions(enemy):
    # returns True when a win condition is met
    if enemy['give_up'][0] >= 100: # enemy gives up peacefully
        text_scroll(f"\n{enemy['give_up'][1]}\n")
        gain_xp(enemy)
        player['spared'] += 1
        return True
    if player['health'] <= 0: # player dies
        text_scroll(f'\n{enemy["name"]} has defeated you...\n')
        game_over()
        return True
    if enemy['health'] <= 0: # player kills enemy
        text_scroll(f'\n{enemy["name"]} has fallen... You Win!\n')
        gain_xp(enemy)
        player['killed'] += 1
        return True
    return False

def check_positive(dict, key, index=0):
    if type(dict[key]) == list:
        if dict[key][index] < 0:
            dict[key][index] = 0
    else:
        if dict[key] < 0:
            dict[key] = 0

def varied_response(response1, response2, pause=0.5):
    text_scroll(f'\n{response1}\n')
    time.sleep(pause)
    text_scroll(f'\n{response2}\n')

def stat_change(entity, stat, value, silent=False):
    entity[stat] += value
    if entity[stat] < 0: 
        entity[stat] = 0
    if not silent:
        if value < 0:
            text_scroll(f'\n{entity["name"]}\'s {stat.upper()} decreased by {abs(value)}!\n')
        else:
            text_scroll(f'\n{entity["name"]}\'s {stat.upper()} increased by {value}!\n')

def configure_params(params, enemy, turns, silent=False):
    # configures parameters for enemy and player items
    updated_params = []
    for par in params:
        if par == '$enemy':
            updated_params.append(enemy)
        elif par == '$turns':
            updated_params.append(turns)
        elif par == '$silent':
            updated_params.append(silent)
        else:
            updated_params.append(par)
    return updated_params

# ---------------------------------------- BATTLE ----------------------------------------

def battle(enemy):
    in_battle = True
    responses = enemy['responses']
    response_keys = list(enemy['responses'].keys())
    enemy_name = enemy['name']

    text_scroll(f"\n{enemy['intro']}\n")
    time.sleep(0.5)
    
    bonus_damage = 0
    bonus_def = 0
    turns = 0
    items_in_use = []

    while in_battle:
    
        if check_win_conditions(enemy):
            return None

        attack = True
        
        

        #choices are below
        print(f'\nYOUR HP: {player["health"]}')
        text_scroll('Would you like to 1) ATTACK, 2) ITEM, 3) ACT, or 4) RUN?\n\nYour choice: ')
        choice = str(input(''))

        # ------------- CHOICE 1: ATTACK -------------

        if choice == '1':
            base_attack = player['attack']
            defense = enemy['defense']
            difference = quicktime_bar('f')
            if difference == 0:
                print('Critical!\t')
                base_damage = base_attack + round(base_attack*0.2)
            elif difference + enemy['speed'] <= 3:
                print('Great Hit!\t')
                base_damage = base_attack
            elif difference + enemy['speed'] <= 7:
                print('Good Hit!\t')
                base_damage = base_attack - math.ceil(base_attack*0.2)
            elif difference + enemy['speed'] <= 12:
                print('Fair Hit!\t')
                base_damage = base_attack - math.ceil(base_attack*0.5)
            else:
                print('Miss!\t\t')
                base_damage = 0
            if base_damage > 0:
                damage = (base_damage + bonus_damage) - defense
                if damage < 0: damage = 0
                enemy['health'] -= damage
                text_scroll(f'\nYou hit {enemy_name} for {damage} damage! {enemy_name} is now at {enemy["health"]} HP!\n')
            else:
                text_scroll('\nYou missed...\n')
            
            turns += 1
            
        # ------------- CHOICE 2: ITEM -------------    

        elif choice == '2':
            inventory = player['inventory']
            if len(inventory) == 0:
                text_scroll('\nYou have no ITEMS!\n')
                attack = False
            else:
                choosing = True
                while choosing:
                    i = 0
                    print('\n0 : EXIT')
                    for item in inventory:
                        print(f'{i + 1} : {item[1].upper()} x{item[4]}')
                        i += 1
                    text_scroll('\nWhat is your choice (enter number)? ')
                    item_choice = str(input(''))
                    if item_choice in set([str(num) for num in range(1, len(inventory) + 1)]):
                        choosing = False
                        cont_item = True
                    elif item_choice == '0':
                        choosing = False
                        cont_item = False
                        attack = False
                
                if cont_item:
                    item_index = int(item_choice) - 1
                    if inventory[item_index][4] <= 1:
                        item = inventory.pop(item_index)
                    else:
                        item = inventory[item_index]
                        inventory[item_index][4] -= 1
                    purpose = str(item[0])
                    name = str(item[1])
                    if type(item[2]) == type(add_item):
                        value = item[2]
                    else:
                        value = int(item[2])
                    if type(item[3]) == int:
                        num_of_turns = item[3]
                    elif type(item[3]) == list:
                        params = item[3]
                    else:
                        dodgable = item[3]
                    turns += 1

                    if purpose == 'bonus attack' or purpose == 'bonus defense':
                        items_in_use.append((purpose, value, turns + num_of_turns, name))           

                    text_scroll(f'\nYou used {name.upper()}!\n')

                    if purpose == 'heal':
                        player['health'] += value
                        if player['health'] > player['max_health']:
                            player['health'] = player['max_health']
                            text_scroll(f'\nYou are now at MAX HP!\n')
                        else:
                            text_scroll(f'\nYou gained {value} HP!\n')
                    elif purpose == 'bonus attack':
                        bonus_damage += int(value)
                        text_scroll(f'\nATTACK increased by {value} for {num_of_turns} turn(s)!\n')
                    elif purpose == 'bonus defense':
                        bonus_def += value
                        text_scroll(f'\nDEFENSE increased by {value} for {num_of_turns} turn(s)!\n')
                    elif purpose == 'damage':
                        chance_num = random.randint(1, 100)
                        if chance_num >= int((enemy['speed'] / 20) * 100) or not dodgable:
                            item_damage = value - enemy['defense']
                            if item_damage < 0: item_damage = 0
                            enemy['health'] -= item_damage
                            text_scroll(f'\nYou did {item_damage} damage to {enemy_name}! {enemy_name} is now at {enemy["health"]} HP!\n')
                        else:
                            text_scroll(f'\n{enemy["name"]} dodged the {name.upper()}!\n')
                    elif purpose == 'special':
                        
                        if '&turns' in params:
                            turns = value(configure_params(params, enemy, turns))
                        else:
                            value(configure_params(params, enemy, turns))

        # ------------- CHOICE 3: ACT -------------

        elif choice == '3':
            text_scroll(f'\n1) {response_keys[0].upper()}, 2) {response_keys[1].upper()}, 3) {response_keys[2].upper()}, 4) {response_keys[3].upper()}, 5) CHECK, or 0) EXIT?\n\nYour choice: ')
            choice = str(input(''))
            random_response = ''
            if choice == '1':
                random_response = random.choice(responses[response_keys[0]])
                varied_response(random_response[0][0], random_response[0][1])
                enemy['give_up'][0] += random_response[1]
                check_positive(enemy, 'give_up', 0)
            elif choice == '2':
                random_response = random.choice(responses[response_keys[1]])
                varied_response(random_response[0][0], random_response[0][1])
                enemy['give_up'][0] += random_response[1]
                check_positive(enemy, 'give_up', 0)
            elif choice == '3':
                random_response = random.choice(responses[response_keys[2]])
                varied_response(random_response[0][0], random_response[0][1])
                enemy['give_up'][0] += random_response[1]
                check_positive(enemy, 'give_up', 0)
            elif choice == '4':
                random_response = random.choice(responses[response_keys[3]])
                varied_response(random_response[0][0], random_response[0][1])
                enemy['give_up'][0] += random_response[1]
                check_positive(enemy, 'give_up', 0)
            elif choice == '5':
                if enemy['checkable'][0]:
                    print(f"\nNAME: {enemy_name}\nATTACK: {enemy['attack']}\nDEFENSE: {enemy['defense']}\nSPEED: {enemy['speed']}\nHEALTH: {enemy['health']}\n")
                    text_scroll(f'{enemy["description"]}\n')
                    time.sleep(1)
                    input('Press enter to continue...')
                else:
                    text_scroll(f'\n{enemy["checkable"][1]}\n')
            else:
                turns -= 1
                attack = False
            if len(random_response) >= 3:
                stat_change(enemy, random_response[2], random_response[3])
            turns += 1
        
        # ------------- CHOICE 4: RUN -------------

        elif choice == '4':
            if enemy['can_flee']:
                text_scroll(f'\nIt appears that {enemy_name} was too much for you...\n')
                player['spared'] += 1
                player['fled'] += 1
                return None
            else:
                text_scroll(f'\nYou can\'t run away from {enemy_name}...\n')
        else:
            text_scroll('\nThat was not a choice.\n')
            attack = False

        if check_win_conditions(enemy):
            return None

        for itm in items_in_use:
            if turns >= itm[2]:
                items_in_use.remove(itm)
                text_scroll(f'Your item, {itm[3].upper()}, has worn off!\n')
                if itm[0] == 'bonus attack':
                    bonus_damage -= itm[1]
                else:
                    bonus_def -= itm[1]


         # ------------- ENEMY ACTION -------------

        Enemy_Item.check_active(turns, enemy)

        if attack:
            # number that determines whether or not the enemy will use an item
            action_factor = random.randint(1, 100000)/1000 # done this way for 3 decimal places of precision

            if action_factor <= enemy['use_item_chance'] and len(enemy['inventory']) > 0: # don't want to use and item if there are no items
                choosing_item = True
                max_loops = 100
                loop_count = 0
                while choosing_item: # looping to make sure enemy doesn't use a 'heal' item when it cant

                    if not loop_count <= max_loops:
                        # this prevents this loop from becoming an infinite loop
                        player_attacked(enemy, bonus_def)
                        break

                    item = random.choice(enemy['inventory'])

                    if item.purpose == 'heal' and enemy['health'] >= enemy["max_health"]:
                        if len(enemy['inventory']) == 1:
                            choosing_item = False
                            player_attacked(enemy, bonus_def)
                        else:
                            loop_count += 1
                            continue
                    else:

                        if '&turns' in item.special_params: # if item is a 'special' function that changes turns
                            turns = item.activate(turns, enemy, bonus_def)
                        else:
                            item.activate(turns, enemy, bonus_def)
                        
                        # checking the number of uses for item
                        if item.uses <= 0:
                            enemy['inventory'].pop(enemy['inventory'].index(item))
                            break

                        choosing_item = False   
            
            else:
                player_attacked(enemy, bonus_def)

player = {
    'name': "Bigg_Milk",
    'attack': 15,
    'defense': 5,
    'health': 40,
    'max_health': 40,
    'base_attack': 15,
    'base_defense': 5,
    'xp': 0,
    'xp_to_next': 40,
    'level': 0,
    'killed': 0,
    'spared': 0,
    'fled': 0,
    'inventory': [
        #[purpose, name, value, turns/dodgable/params, uses]
        #purposes are 'heal', 'bonus attack', 'bonus defense', 'damage', 'special'
        ['heal', 'Bandage', 10, None, 2],
        ['bonus attack', 'Energy Drink', 10, 3, 3],
        ['bonus defense', 'Frail Shield', 8, 2, 1, 1],
    ]
}

def save(stage):
    save_file = shelve.open('saves/save.db')
    save_file['player_object'] = player
    save_file['stage'] = stage
    save_file.close()
    

def get_save():
    global stage, player
    with open('saves/save.db.dir', 'r') as sv:
        if len(sv.readlines()) > 1:
            save_file = shelve.open('saves/save.db')
            stage = int(save_file['stage'])
            player = dict(save_file['player_object'])
            save_file.close
        else:
            save(0)
            get_save()

def erase_save():
    with open('saves/save.db.dir', 'w') as file1:
        file1.write('')
    with open('saves/save.db.dat', 'w') as data:
        data.write('')
    

#get_save()
attack_bar = progbar.ProgressBar(bar_size=30)
